### PR TITLE
feat(trainer): Support NPU labels in TrainJob device

### PR DESCRIPTION
**Pull request by sujalshah-bit**
**What this PR does / why we need it**:
Add support for NPU resource labels in container resource limits,
resolving the existing TODO to support additional accelerator types
(e.g., NPUs).

Fixes NONE

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

### CODE REVIEW

File: kubeflow/trainer/backends/kubernetes/utils.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def get_container_devices(
     # TODO (andreyvelich): We should discuss how to get container device type.
     # Potentially, we can use the trainer.kubeflow.org/device label from the runtime or
     # node types.
-    # TODO (andreyvelich): Support other resource labels (e.g. NPUs).
+    # TODO (andreyvelich): Support other resource labels.
```
Comment (andreyvelich) [inline kubeflow/trainer/backends/kubernetes/utils.py]: You can remove this TODO statement.


```suggestion
```

File: kubeflow/trainer/backends/kubernetes/utils.py [hunk 2]
```diff
@@ -55,6 +55,13 @@ def get_container_devices(
         mig_key = mig_keys[0]
         device = mig_key.split("/")[1]
         device_count = resources.limits[mig_key].actual_instance
+    elif npu_keys := [k for k in resources.limits if k.endswith(constants.NPU_LABEL_SUFFIX)]:
+        if len(npu_keys) > 1:
+            raise ValueError(f"Multiple NPU resource types are not supported yet: {npu_keys}")
```
Comment (andreyvelich) [inline kubeflow/trainer/backends/kubernetes/utils.py]: Do we need this?

File: kubeflow/trainer/backends/kubernetes/utils.py [hunk 3]
```diff
@@ -55,6 +55,13 @@ def get_container_devices(
...
+
+        npu_key = npu_keys[0]
```
Comment (andreyvelich) [inline kubeflow/trainer/backends/kubernetes/utils.py]: Why are you taking the first item? I thought NPU resources are set similarly to TPU and GPU

File: kubeflow/trainer/backends/kubernetes/utils_test.py [hunk 4]
```diff
@@ -61,6 +61,31 @@ def _build_runtime() -> types.Runtime:
             },
             expected_error=ValueError,
         ),
+        TestCase(
+            name="single NPU limit returns device and count",
+            expected_status=SUCCESS,
+            config={
+                "resources": models.IoK8sApiCoreV1ResourceRequirements(
+                    limits={
+                        "huawei.com/npu": models.IoK8sApimachineryPkgApiResourceQuantity(2),
```
Comment (andreyvelich) [inline kubeflow/trainer/backends/kubernetes/utils_test.py]: Remove vendors from unit tests

```suggestion
                        "example.com/npu": models.IoK8sApimachineryPkgApiResourceQuantity(2),
```

File: kubeflow/trainer/backends/kubernetes/utils_test.py [hunk 5]
```diff
@@ -61,6 +61,31 @@ def _build_runtime() -> types.Runtime:
...
+                    }
+                )
+            },
+            expected_output=("npu", "2.0"),
+        ),
+        TestCase(
+            name="multiple NPU resource types are not supported",
```
Comment (andreyvelich) [inline kubeflow/trainer/backends/kubernetes/utils_test.py]: This should not be the case.

File: kubeflow/trainer/constants/constants.py [hunk 6]
```diff
@@ -104,6 +104,9 @@
 # The label for TPU in the container resources.
 TPU_LABEL = "google.com/tpu"
 
+# The Suffix label for NPU in the container resources.
+NPU_LABEL_SUFFIX = "/npu"
```
Comment (andreyvelich) [inline kubeflow/trainer/constants/constants.py]: I thought the resource name is: huawei.com/Ascend910.
@danish9039 Do you know details about it?
Ref: https://github.com/kubeflow/sdk/pull/264

File: kubeflow/trainer/backends/kubernetes/utils_test.py (hunk 4 above)
Comment (sujalshah-bit) [reply kubeflow/trainer/backends/kubernetes/utils_test.py]: I understand the preference for vendor-neutral tests, but the current NPU implementation is tied to a specific vendor label (`huawei.com/Ascend910`) similar to how GPU is tied to nvidia.com/gpu and TPU to `google.com/tpu`. 
If we want `example.com/npu` in the test, we'd need to change the implementation to use suffix matching (`/npu`) instead of an exact vendor label, but as discussed, that's not guaranteed to work for all NPU vendors (e.g. huawei.com/Ascend910 has no /npu suffix). Happy to go either way, just need clarity on which approach we're taking.

Comment (andreyvelich) [reply kubeflow/trainer/backends/kubernetes/utils_test.py]: I see, good point. Let's keep it as NPU label for now

**Review by andreyvelich — COMMENTED** [review]
I am fine with starting supporting NPU via `huawei.com/Ascend910` label.
/lgtm
/cc @astefanutti @akshaychitneni @kramaranya @Fiona-Waters

File: kubeflow/trainer/constants/constants.py [hunk 7]
```diff
@@ -104,6 +104,9 @@
 # The label for TPU in the container resources.
 TPU_LABEL = "google.com/tpu"
 
+# The label for NPU in the container resources.
+NPU_LABEL = "/npu"
+
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/constants/constants.py:112]: I think we should be a little clearer in comments as to why we are hardcoding this.

File: kubeflow/trainer/backends/kubernetes/utils.py [hunk 8]
```diff
@@ -55,6 +54,10 @@ def get_container_devices(
         mig_key = mig_keys[0]
         device = mig_key.split("/")[1]
         device_count = resources.limits[mig_key].actual_instance
+    elif constants.NPU_LABEL in resources.limits:
+        # Can't derive cleanly from key unlike gpu/tpu
```
Comment (Fiona-Waters) [inline kubeflow/trainer/backends/kubernetes/utils.py]: Could you? update the coment to explain what the device label is and why we return npu?

File: kubeflow/trainer/constants/constants.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreyvelich) [reply kubeflow/trainer/constants/constants.py:112]: Maybe @JesseStutler @lowang-bh can help us to find the better alias for Huawei's Ascend chips?
Do you know what is the name users usually say when they want to use this AI chip?

[the remark this answers is not in the document]
Comment (JesseStutler) [reply kubeflow/trainer/constants/constants.py:112]: NPU isn't limited to the 910 series chips, it's definitely a bit hard-coded here. However, I'm not familar with the context of this file. If a user wants to use the NPU, and it's a 310 series chip, the they'll request will be like `huawei.com/Ascend310P`. So only the prefix `huawei.com/` is the same, unlike with GPUs. The suffix might vary depending on the specific chip.

[the remark this answers is not in the document]
Comment (andreyvelich) [reply kubeflow/trainer/constants/constants.py:112]: Would it be better to ask users provide the dedicated chip model? For example:
```py
trainer=CustomTrainer(
        func=torch_dist,
        num_nodes=3,
        resources_per_node={"Ascend310P": 2},
    ),

trainer=CustomTrainer(
        func=torch_dist,
        num_nodes=3,
        resources_per_node={"Ascend910C": 5},
    ),
```

We would just add the prefix there: `huawei.com/`

[the remark this answers is not in the document]
Comment (JesseStutler) [reply kubeflow/trainer/constants/constants.py:112]: I prefer this style as @andreyvelich mentioned

[the remark this answers is not in the document]
Comment (andreyvelich) [reply kubeflow/trainer/constants/constants.py:112]: @sujalshah-bit please can you update the PR to let users set the `Ascend***` device?

**Comment by andreyvelich** [comment]
@sujalshah-bit Please could you address @Fiona-Waters comment, so we can move this forward?

**Comment by andreyvelich** [comment]
Ping @sujalshah-bit.
Did you get a chance to address the small comments, so we can merge this PR?

**Comment by sujalshah-bit** [comment]
@andreyvelich, sorry for the delay and any inconvenience. I keep forgetting about this. Could you please review the latest change :)

**Review by andreyvelich — COMMENTED** [review]
Thanks @sujalshah-bit!
/ok-to-test
/lgtm
/approve
